### PR TITLE
feat: Digitization 2, introduce SurfaceMasker

### PR DESCRIPTION
**Pull request by asalzburger**
This PR is the second step of the new Digitisation.

It introduces the `SurfaceMasker` that is run before the `Channelizer` which was introduced in part 1.
The `PlanarSurfaceMasker` clips/masks a segment onto the surface (= readout) bounds.

Additionally, it harmonizes the input/output between Channelized / Masker.

It also reorders the PlanarSurface testbeds to be used for `ChannelizerTest` and `PlanarSurfaceMaskTest`.

* `RectangleBounds`
- green : inside from the the start (untouched)
- blue: clipped

Unknown-2

* `TrapezoidBounds`

Unknown

* `RadialBounds`

- ref: segment fully outside

Unknown-3

### CODE REVIEW

**Comment by asalzburger** [comment]
Hi @XiaocongAi - this should answer your question about outside segments.

File: Tests/UnitTests/Fatras/Digitization/PlanarSurfaceMaskTests.cpp
Comment (XiaocongAi) [inline Tests/UnitTests/Fatras/Digitization/PlanarSurfaceMaskTests.cpp:114]:
```suggestion
  /// Case five: outside phi max
```

**Review by XiaocongAi — COMMENTED** [review]
Hi @asalzburger , it looks good to me. There is only the Annulus edge vector I'm not very sure (But since there is the unit test for this, I guess you should be right).

File: Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp
Comment (asalzburger) [reply Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:32]: Fixed.

File: Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp
Comment (asalzburger) [reply Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:46]: Fixed.

File: Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp
Comment (asalzburger) [reply Fatras/include/ActsFatras/Digitization/PlanarSurfaceMask.hpp:42]: Fixed.

File: Fatras/src/Digitization/DigitizationError.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/DigitizationError.cpp:26]: Fixed.

File: Tests/UnitTests/Fatras/Digitization/DigitizationCsvOutput.hpp
Comment (asalzburger) [reply Tests/UnitTests/Fatras/Digitization/DigitizationCsvOutput.hpp:35]: Fixed/

File: Tests/UnitTests/Fatras/Digitization/PlanarSurfaceMaskTests.cpp
Comment (asalzburger) [reply Tests/UnitTests/Fatras/Digitization/PlanarSurfaceMaskTests.cpp:114]: Solved.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:153]: I will check.

**Comment by asalzburger** [comment]
There was a problem/bug in the `DiscTrapezoidBounds`, which was only partly fixed in the `Digitization 1` PR, this fixes it completely now.

**Comment by asalzburger** [comment]
Unknown

This is for `DiscTrapezoidBounds`.

**Comment by asalzburger** [comment]
Still investigating the `AnnulusBounds` issue, so changing to `WIP` now.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp]: I rewrote that part.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp]: Again, I rewrote that part.

**Comment by asalzburger** [comment]
Unknown-1

Now with fixed unit test & bug fix in AnnulusBounds.

File: Core/include/Acts/Surfaces/AnnulusBounds.hpp
Comment (asalzburger) [inline Core/include/Acts/Surfaces/AnnulusBounds.hpp:157]: This was the bug mentioned, we used undefined `m_phiAvg` which resulted in useless results.

**Comment by asalzburger** [comment]
Hi @XiaocongAi  - this is now updated, I shuffled the code around, but it's largely the same. Also, I fixed the problem with the `AnnulusBounds` which had an internal bug.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (XiaocongAi) [inline Core/src/Surfaces/AnnulusBounds.cpp:107]: Could it be that `phiMinInner` is derived from `m_inRightStripXY` and `phiMaxInner` derived from `m_inLeftStripXY`?

**Review by XiaocongAi — COMMENTED** [review]
Hi @asalzburger , the updated approach to get the segment intersection looks more clear. But there are some points I'm not so sure.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (XiaocongAi) [inline Core/src/Surfaces/AnnulusBounds.cpp:117]: Since it's starting from the left inner corner, it's `phi_max->phi_min`?

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:232]: Hi @XiaocongAi - you are good. Indeed, this is the case that could happen, BUT ... This masker will operate on hits that are on surface and the reason why they could be outside of the mask is purely due to the incident angle and the charge drift. 

Both makes it practically impossible to create a case you describe.

As > 99 % of all cases are safe with both inside and the masker will not even operate, I do not want to penalise the 99% good cases with extra checks, that are pathological.

I will clarify this in the description.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (asalzburger) [reply Core/src/Surfaces/AnnulusBounds.cpp:107]: Left, right - that's so complicated :-P

I will fix this.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (asalzburger) [reply Core/src/Surfaces/AnnulusBounds.cpp:117]: Will correct with left/right correctly.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (XiaocongAi) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:232]: Yeah, from the realistic point of view, we are indeed safe.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:39]: Fixed.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:41]: Fixed.

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:221]: I know, the idea is the following:
- if any of them it outside : create a 2D intersection (which one is found out later).

File: Fatras/src/Digitization/PlanarSurfaceMask.cpp
Comment (asalzburger) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:232]: This is what I added:

```c++
  /// @note If both end points of the segment are inside, the segment
  /// is not clipped/masked, even if it would cross a surface boundary.
  /// Examples for those would be non-covex polygons or segments on a
  /// radial bound, where the radial boundary is crossed. Such segments
  /// do not occur in Digitization, as the hit has to be inside the 
  /// surface bounds to start with.
```

Comment (XiaocongAi) [reply Fatras/src/Digitization/PlanarSurfaceMask.cpp:232]: Great! :+1:

**Review by asalzburger — COMMENTED** [review]
Addressed all comments.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (asalzburger) [reply Core/src/Surfaces/AnnulusBounds.cpp]: Done.

File: Core/src/Surfaces/AnnulusBounds.cpp
Comment (asalzburger) [reply Core/src/Surfaces/AnnulusBounds.cpp:114]: It actually worked by accident (at least for the obj displayer), because it has to create a full closed face.
Fixed now.

**Review by XiaocongAi — APPROVED** [review]
Thanks! Approve.

**Comment by asalzburger** [comment]
Waiting for CI to complete before merging.

**Pull request merged**
